src/audio/dai.c

```diff
@@ -637,6 +713,62 @@ static int dai_comp_trigger(struct comp_dev *dev, int cmd)
 	return ret;
 }
 
+static int dai_comp_trigger(struct comp_dev *dev, int cmd)
+{
+	struct dai_data *dd = comp_get_drvdata(dev);
+	struct dai_group *group = dd->group;
+	uint32_t irq_flags;
+	int ret = 0;
+
+	if (!group) {
+		comp_dbg(dev, "dai_comp_trigger(), non-atomic trigger");
+		return dai_comp_trigger_int(dev, cmd);
+	}
+
+	if (group->trigger_cmd < 0) {
+		comp_dbg(dev, "dai_comp_trigger(), begin atomic trigger for group %d",
+			 group->group_id);
+		group->trigger_cmd = cmd;
+		group->trigger_counter = group->num_dais - 1;
+	} else if (group->trigger_cmd != cmd) {
+		comp_err(dev, "dai_comp_trigger(), already processing atomic trigger");
+		/* already processing a different trigger command */
+		ret = -EAGAIN;
+	} else {
+		group->trigger_counter--;
+		comp_dbg(dev, "dai_comp_trigger(), trigger counter %d, group %d",
+			 group->trigger_counter, group->group_id);
+
+		if (!group->trigger_counter) {
+			/*
+			 * The counter has reached 0, which means
+			 * all DAIs have received the same trigger command
+			 * and we may begin the actual trigger process
+			 * synchronously.
+			 */
+
+			irq_local_disable(irq_flags);
+			notifier_event(group, NOTIFIER_ID_DAI_TRIGGER,
+				       BIT(cpu_get_id()), NULL, 0);
+			irq_local_enable(irq_flags);
+
+			ret = group->trigger_ret;
+			group->trigger_cmd = -1; /* no trigger command */
+		}
+	}
+
+	return ret;
```

Comment:
I see you toy around with the assumption that trigger commands return 0 on success and <0 on failure. Valid assumption, nevertheless it's good to have it documented.

Reply:
I don't quite understand what you mean.
I'm merely forwarding the return value of last trigger operation.

Reply:
Well in the cores that don't quite trigger you just count the number of triggers and return 0. If the trigger operation itself returns something nonzero on success from DAIs (I'm not sure this is actually the case) then you may break the interface for those that actually use the return value further than the mere error-checking. The fact that you only report error for one of the calls is fine though, because that error will tear down everything anyway and it doesn't matter exactly where it happened.